### PR TITLE
Replace Reflection with Factory Function for Filter Instantiation

### PR DESCRIPTION
**Pull request by commoddity**
Hey @ardikabs . First of all, love this repo. It makes writing Go filters for Envoy much easier; thanks for creating it.

The reasoning for this PR is that I ran into an issue where I needed to inject a dependency into an HTTPFilter struct and noticed that your usage of `Reflect` was making this not possible due to it creating a new copy of the Filter on every request. This PR is proposing an alternate method that uses a factory function to create the `Filter`, which allows passing dependencies into the filter that will be preserved. 

For example, in my usage I am passing a pointer to a cache that the filter can then access:

```go
filterFactory := func() gonvoy.HttpFilter {
    return &Filter{cache: cache}
}

gonvoy.RunHttpFilter(filterName, filterFactory, gonvoy.ConfigOptions{
    FilterConfig:            new(Config),
    DisableStrictBodyAccess: true,
})
```

I also replaced the `Name()` method with a string argument as the filter is not yet instantiated at the point the factory func is passed to `RunHttpFilter`.

Here's a little more summary on the reasoning and advantages:

**Summary:**

This PR refactors the filter instantiation process by replacing reflection with a factory function. This change improves performance, enhances type safety, and simplifies the codebase.

**Advantages:**

- **Performance**: Eliminates reflection overhead for faster execution.
- **Type Safety**: Enables compile-time checks to prevent runtime errors.
- **Code Clarity**: Simplifies code, making it easier to read and maintain.
- **Dependency Injection**: Allows explicit injection of dependencies into filters.
	
Let me know what you think and thanks again for the repo.

### CODE REVIEW

**Comment by ardikabs** [comment]
i like this idea, will review it soon.
Anyway thanks for the PR @commoddity !

File: test/e2e/filters/http_headers_modifier/main.go [hunk 1]
```diff
@@ -6,22 +6,32 @@ import (
...
+		func() gonvoy.HttpFilter {
+			return new(Filter)
+		},
+		gonvoy.ConfigOptions{
+			FilterConfig: new(Config),
+		},
+	)
+
+	gonvoy.RunHttpFilter(
+		filterName,
```
Comment (ardikabs) [inline test/e2e/filters/http_headers_modifier/main.go]: this e2e tests use 2 different filters, the `http_headers_modifier` and `echoserver`, hence it shouldn't reuse the `filterName`.

File: test/e2e/filters/http_body/main.go [hunk 2]
```diff
@@ -5,26 +5,41 @@ import (
 )
 
 func init() {
-	gonvoy.RunHttpFilter(new(BodyReadFilter), gonvoy.ConfigOptions{
-		FilterConfig: new(Config),
+	gonvoy.RunHttpFilter(filterName,
```
Comment (ardikabs) [inline test/e2e/filters/http_body/main.go]: similarly on this one, `filterName` must differ on every filters being registered

File: test/e2e/filters/http_body/main.go [hunk 3]
```diff
@@ -5,26 +5,41 @@ import (
...
-	})
+			DisableStrictBodyAccess: true,
+			EnableRequestBodyRead:   true,
+			EnableResponseBodyRead:  true,
+		},
+	)
 
-	gonvoy.RunHttpFilter(new(BodyWriteFilter), gonvoy.ConfigOptions{
-		FilterConfig: new(Config),
+	gonvoy.RunHttpFilter(filterName,
```
Comment (ardikabs) [inline test/e2e/filters/http_body/main.go]: ditto

File: test/e2e/filters/http_body/main.go [hunk 4]
```diff
@@ -5,26 +5,41 @@ import (
...
+			EnableRequestBodyWrite:  true,
+			EnableResponseBodyWrite: true,
+		},
+	)
 
-	gonvoy.RunHttpFilter(new(Echoserver), gonvoy.ConfigOptions{
-		DisableStrictBodyAccess: true,
-		EnableRequestBodyRead:   true,
-	})
+	gonvoy.RunHttpFilter(filterName,
```
Comment (ardikabs) [inline test/e2e/filters/http_body/main.go]: ditto

File: httpfilter.go [hunk 5]
```diff
@@ -16,14 +16,14 @@ var NoOpHttpFilter = &api.PassThroughStreamFilter{}
 //
 //	package main
 //	func init() {
-//		RunHttpFilter(new(UserhttpFilter), ConfigOptions{
+//		RunHttpFilter(filterName, filterFactory, ConfigOptions{
 //			BaseConfig: new(UserHttpFilterConfig),
 //		})
 //	}
-func RunHttpFilter(filter HttpFilter, options ConfigOptions) {
+func RunHttpFilter(filterName string, filterFactoryFunc func() HttpFilter, options ConfigOptions) {
```
Comment (ardikabs) [inline httpfilter.go]: i suggest to introduce a new type instead of `func() HttpFilter`, like `HttpFilterFactory`, wdyt @commoddity ?

**Review by ardikabs — CHANGES_REQUESTED** [review]

File: test/e2e/filters/http_headers_modifier/main.go (hunk 1 above)
Comment (commoddity) [reply test/e2e/filters/http_headers_modifier/main.go]: Sounds good to me. I will update the test files when back at home today or tomorrow.

File: httpfilter.go (hunk 5 above)
Comment (commoddity) [reply httpfilter.go]: That sounds like a good idea. I will implement your suggestions and update the PR later today or tomorrow. Thanks for the quick review.

Comment (commoddity) [reply httpfilter.go]: Have updated to address this and the other review comments.

**Comment by ardikabs** [comment]
LGTM! Thanks for the PR!

**Pull request merged**